src/Microsoft.AspNetCore.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -88,11 +90,25 @@ public MediaType(string mediaType, int offset, int? length)
             if (subTypeLength == 0)
             {
                 SubType = new StringSegment();
+                SubTypeWithoutSuffix = new StringSegment();
+                SubTypeSuffix = new StringSegment();
                 return;
             }
             else
             {
                 SubType = subType;
+                
+                var subtypeWithoutSuffixLength = TryGetSuffixLength(subType, out var subtypeSuffixLength)
+                    ? subType.Length - subtypeSuffixLength - 1
+                    : subType.Length;
+
+                SubTypeWithoutSuffix = subtypeWithoutSuffixLength > 0
+                    ? subType.Subsegment(0, subtypeWithoutSuffixLength)
+                    : new StringSegment();
+
+                SubTypeSuffix = subtypeSuffixLength > 0
+                    ? subType.Subsegment(subtypeWithoutSuffixLength + 1, subtypeSuffixLength)
+                    : new StringSegment();
```

Comment:
I did it this way to degenerate inputs like `application/+json` and `application/entity+`, making sure that `SubtypeWithoutSuffix` and `SubtypeSuffix` (respectively) have `HasValue`=false in these two degenerate cases. However, on further consideration, it's probably better *not* to behave as if `application/entity+` is the same as `application/entity` (etc.).

So I've now simplified this down to a single `if` like you suggest, meaning that `application/entity` now has its `.SubtypeSuffix.HasValue == true`, whereas `application/entity+` results in `.SubTypeSuffix.HasValue == false` (it's a zero-length substring, rather than a no-value substring).